### ISSUE TITLE
Is there any demo for running java + nodejs in one vm?

### ISSUE THREAD

**Issue by balayanv** [graalvm/graalvm-demos#151]
I simply want to create a POC with java using nodejs to utilise some npm packages.
Being able to submit async tasks to node is part of the goal.

**Comment by olyagpl** [comment]
You cannot really use any npm module from Java "as is". You have to apply a package bundler to create a "pure JS" version of that module. Then it should be possible to call it from Java  (assuming the package does not have dependencies on node.js modules like `fs`). Given that there are multiple package bundlers and packages may change over time, we advise users to use their own favorite package bundler for their own application scenario such as webpack/browserify/etc. This is essentially what we advise in [our documentation](https://www.graalvm.org/latest/reference-manual/js/JavaInteroperability/#polyglot-context):

Supported NPM packages can be used in a GraalVM JavaScript Context using one of the following approaches:
Using a package bundler. For example, to combine multiple NPM packages in a single JavaScript [Source file](https://www.graalvm.org/sdk/javadoc/org/graalvm/polyglot/Source.html).
Using ES modules on the local FileSystem. Optionally, a custom [Truffle FileSystem](https://www.graalvm.org/truffle/javadoc/org/graalvm/polyglot/io/FileSystem.html) can be used to configure how files are resolved.

Some popular npm packages provide a "pure JS" version through some CDNs, for example, https://cdnjs.cloudflare.com/ajax/libs/validator/13.5.2/validator.min.js.

You can have a look at the MLE demo, https://github.com/stefandobre/apex-mle-demo,  running an npm package in an MLE instance is almost equivalent to running from a Java context.

**Comment by balayanv** [comment]
I have got it working, but there is much working around that i had to do to achieve it.
Imo no one is going to use all this great work you put into this, if you dont do the last step and make it straight forward to use.

I had to essentially run it as node application that calls a java class that initializes a threadsafe queue for putting tasks for node loop to execute... its very annoying to use since now you cant debug it from idea, you have to run it with custom executable that pretends to be java.

Its a bad dev experience. Problem here is that almost everything thats useful depends on your javascript environment either being in a browser or node, mostly for doing simple http and websocket stuff. 

Either need to provide transparent substitutions for this or make it such that we run pure node but with java together in one vm. Thats my 2 cents...

**Comment by balayanv** [comment]
You can hardly find any npm modules that do not depend on some of the essential things like http or ws.

**Comment by balayanv** [comment]
I'd suggest you to attack the problem from this angle. Why would java devs be interested in graal?

Mostly to make use of whats in js/node communit without rewriting it in java, or python community without rewriting it in java + optioanlly potentially have higher performance since all that will be chewed by JIT into bytecode that gets optimised into high performance code.

If you agree that this is a goal : then focus on some minimal prototype that can use all that graalvm already has to provide this to java devs. What you have is impressive but not that practical.

I see lots of energy put into things like AOT native images... who cares about native images? who cares about startup time when you trade performance for it. Only real place where this might matter is some sort of AWS lambda environment. Even then its an overkill.

Basically if java devs want to use node/js/python existing libraries its still a pain and you essentially run node or python and hack java into it.
